calico/openstack/mech_calico.py

```diff
@@ -196,14 +205,163 @@ def _init_state(self):
+
+            # Only handle updates if we are the master node.
+            if self.transport.is_master:
+                LOG.info("I am master: polling felix updates in etcd")
+                self._poll_and_handle_felix_updates(expected_epoch)
+            else:
+                # Short sleep interval before we check if we've become
+                # the master.
+                eventlet.sleep(MASTER_CHECK_INTERVAL_SECS)
+        else:
+            LOG.warning("Unexpected fork. "
+                        "Handling status updates thread exiting.")
+
+
+    def _poll_and_handle_felix_updates(self, expected_epoch):
+        """
+        In infinite loop read new statuses in etcd status directory. Then
+        pass updates to Neutron database. On exception reconnect the etcd
+        client and do complete resync when necessary.
+        """
+        try:
+            while True:
+                # Get and handle an update.
+                LOG.debug("Reading from etcd. Dir: " + STATUS_DIR + "index: " + str(self.next_etcd_index))
+                response = self._client.read(STATUS_DIR,
+                                             wait=True,
+                                             waitIndex = self.next_etcd_index,
+                                             recursive=True)
+
+                # Before status is passed to Neutron db, check whether we
+                # are still master (if not, we loop until we become one)
+                # and whether epoch is as expected (if not, thread exits).
+                if self._epoch != expected_epoch or not self.transport.is_master:
+                    break
+
+                self._handle_status_update(response)
+                # Since we're polling on a subtree, we can't just
+                # increment the index, we have to look at the
+                # modifiedIndex.
+                self.next_etcd_index = max(self.next_etcd_index,
+                                           int(response.modifiedIndex)) + 1
+        except (ReadTimeoutError, SocketTimeout) as e:
+            # This is expected when we're doing a poll and nothing
+            # happened. socket timeout doesn't seem to be caught by
+            # urllib3 1.7.1. Simply reconnect.
+            LOG.debug("Read from etcd timed out (%r), retrying.", e)
+            self._client = etcd.Client(host=cfg.CONF.calico.etcd_host,
+                                       port=cfg.CONF.calico.etcd_port)
+        except (etcd.EtcdClusterIdChanged, etcd.EtcdEventIndexCleared) as e:
+            # Complete resync is needed.
+            LOG.warning("Out of sync with etcd (%r).  Reconnecting "
+                     "for full sync.", e)
+            try:
+                # Connect the client
+                self._client = etcd.Client(host=cfg.CONF.calico.etcd_host,
+                                           port=cfg.CONF.calico.etcd_port)
+                context = ctx.get_admin_context()
```

Comment:
The next few lines of code should be refactored into a method, which can be called both from here, and from the periodic resync thread. Speaking of which...

What happens if the resync thread gets scheduled while this thread is resyncing? The answer might be that everything is fine, but if the two threads can trip each other up, then you will need to have some kind of flag saying 'resync in progress', to stop the two threads interfering with each other.